src/Logging.W3C/src/W3CLoggerProcessor.cs

```diff
@@ -0,0 +1,195 @@
+using System;
+using System.Collections.Concurrent;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Extensions.Logging.W3C
+{
+    internal class W3CLoggerProcessor : IDisposable
+    {
+        private const int _maxQueuedMessages = 1024;
+
+        private readonly string _path;
+        private readonly string _fileName;
+        private readonly int? _maxFileSize;
+        private readonly W3CLoggingFields _loggingFields;
+        private bool _hasWritten;
+
+        private readonly BlockingCollection<string> _messageQueue = new BlockingCollection<string>(_maxQueuedMessages);
+        private Task _outputTask;
+
+        public W3CLoggerProcessor(IOptionsMonitor<W3CLoggerOptions> options)
+        {
+            var loggerOptions = options.CurrentValue;
+            _path = loggerOptions.LogDirectory;
+            _fileName = loggerOptions.FileName;
+            _maxFileSize = loggerOptions.FileSizeLimit;
+            _loggingFields = loggerOptions.LoggingFields;
+
+            FullName = Path.Combine(_path, $"{_fileName}{Guid.NewGuid().ToString()}.log");
+
+            // Start W3C message queue processor
+            _outputTask = Task.Run(ProcessLogQueue);
+        }
+
+        public string FullName { get; }
+
+        public void EnqueueMessage(string message)
+        {
+            // Write log directives the first time we log a message
+            if (!_hasWritten)
+            {
+                WriteDirectives();
+            }
+            if (!_messageQueue.IsAddingCompleted)
+            {
+                try
+                {
+                    _messageQueue.Add(message);
+                    return;
+                }
+                catch (InvalidOperationException) { }
+            }
+        }
+
+        internal async Task WriteMessageAsync(string entry, StreamWriter streamWriter)
+        {
+            await streamWriter.WriteLineAsync(entry);
+            await streamWriter.FlushAsync();
+        }
+
+        private async Task ProcessLogQueue()
+        {
+            try
+            {
+                Directory.CreateDirectory(_path);
+                var fileInfo = new FileInfo(FullName);
+                using (var streamWriter = File.AppendText(FullName))
+                {
+                    foreach (string message in _messageQueue.GetConsumingEnumerable())
+                    {
+                        // Only write until we've reached _maxFileSize
+                        if (fileInfo.Exists && fileInfo.Length < _maxFileSize)
```

Comment:
TODO - should I create a new file when _maxFileSize is reached? Right now I just stop writing.

Reply:
Should roll like Logging.AzureAppServices